src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,916 @@
+/**
+ * @file feature_extraction_Impl.hpp
+ * @author Nilay Jain
+ *
+ * Implementation of feature extraction methods.
+ */
+#ifndef MLPACK_METHODS_EDGE_BOXES_STRUCTURED_TREE_IMPL_HPP
+#define MLPACK_METHODS_EDGE_BOXES_STRUCTURED_TREE_IMPL_HPP
+
+
+#include "feature_extraction.hpp"
+#include <mlpack/methods/pca/pca.hpp>
+
+namespace mlpack {
+namespace structured_tree {
+
+
+template<typename MatType, typename CubeType>
+StructuredForests<MatType, CubeType>::
+StructuredForests(mlpack::structured_tree::FeatureParameters F) : params(F) {}
+
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+GetFeatureDimension(arma::vec& FtrDim)
+{
+  FtrDim = arma::vec(2);
+
+  const size_t shrink = this->params.Shrink();
+  const size_t pSize = this->params.PSize();
+  const size_t numCell = this->params.NumCell();
+  const size_t numOrient = this->params.NumOrient();
+  const size_t nColorCh = params.RGBD() == 0 ? 3 : 4;
+  const size_t nCh = nColorCh + 2 * (1 + numOrient);
+  FtrDim[0] = std::pow((pSize / shrink) , 2) * nCh;
+  FtrDim[1] = std::pow(numCell , 2) * (std::pow (numCell, 2) - 1) / 2 * nCh;
+}
+
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+DistanceTransform1D(const arma::vec& f, const size_t n, const double inf,
+                    arma::vec& d)
+{
+  arma::vec v(n), z(n + 1);
+  d = arma::vec(n);
+  size_t k = 0;
+  v[0] = 0.0;
+  z[0] = -inf;
+  z[1] = inf;
+  for (size_t q = 1; q <= n - 1; ++q)
+  {
+    float s  = ( (f[q] + q * q)-( f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    while (s <= z[k])
+    {
+      --k;
```

Comment:
I'm not sure, but if we set k = 0 at the beginning, how can we decrease k (``size_t``), and still get the right index?